### PR TITLE
scripts: xtensa-build-all.sh: Refactor and improve script

### PR DESCRIPTION
**Pull request by crazoes**
- Use getopts for simpler, robust argument parsing
- Exit if an unknown platform is specified
- Derive $ROOT from $HOST instead of explicit assignment
- Use switch case instead of if else for testing string

Signed-off-by: Shreeya Patel <shreeya.patel23498@gmail.com>

### CODE REVIEW

**Comment by crazoes** [comment]
Furthers https://github.com/thesofproject/sof/issues/2631

**Comment by crazoes** [comment]
Jenkins probably uses `./scripts/xtensa-build-all.sh -j $PLATFORM` as well, although can't seem to find any jenkinsfile in the repo.

File: .travis.yml [hunk 1]
```diff
@@ -31,7 +31,7 @@ env:
   - PLATFORM=imx8x
   - PLATFORM=imx8m
 
-script: ./scripts/docker-run.sh ./scripts/xtensa-build-all.sh -j $PLATFORM
+script: ./scripts/docker-run.sh ./scripts/xtensa-build-all.sh $PLATFORM
```
Comment (lgirdwood) [inline .travis.yml:34]: Sorry, not following the commit message. Why do we not want the -j option here ?

Comment (crazoes) [reply .travis.yml:34]: Because `-j` option has to provide the number of cores to use. By default, with or without the option and option's value all the cores are used. So this option is not adding any value here. And, with my update, `getopts` enforces that `-j` requires a value with it.

Hence, removing it makes use of all cores like earlier, not altering the CI.

**Comment by lgirdwood** [comment]
SOFCI TEST

File: scripts/xtensa-build-all.sh [hunk 2]
```diff
@@ -8,87 +8,64 @@ BUILD_ROM=no
 BUILD_DEBUG=no
 BUILD_FORCE_UP=no
```
Comment (dbaluta) [inline scripts/xtensa-build-all.sh:9]: @crazoes change looks good to me. Anyhow, for easier review can you please split this patch so that each commit only does one thing!

I think each of your list item in commit description should be a separate commit. Commits should be as separate as possible, and they should work independently of following commits in order to not break git bisection.

Comment (crazoes) [reply scripts/xtensa-build-all.sh:9]: @dbaluta i have split the commit

File: scripts/xtensa-build-all.sh [hunk 3]
```diff
@@ -113,171 +90,148 @@ set -e
...
+			XTENSA_TOOLS_VERSION="RG-2017.8-linux"
+			;;
+		apl)
+			PLATFORM="apollolake"
+			ARCH="xtensa-smp"
+			XTENSA_CORE="X4H3I16w2D48w3a_2017_8"
+
+			# test APL compiler aliases and ignore set -e here
+			type xtensa-bxt-elf-gcc > /dev/null 2>&1 && true
+			if [ $? == 0 ]
```
Comment (marc-hb) [inline scripts/xtensa-build-all.sh:135]:
```
if type -p xtensa-bxt-elf-gcc ; then
  ...
```
shellcheck should flag this.

The redirections don't seem useful to me.

Same below.

File: scripts/xtensa-build-all.sh [hunk 4]
```diff
@@ -113,171 +90,148 @@ set -e
...
+			HOST="xtensa-hsw-elf"
+			XTENSA_TOOLS_VERSION="RG-2017.8-linux"
+			;;
+		apl)
+			PLATFORM="apollolake"
+			ARCH="xtensa-smp"
+			XTENSA_CORE="X4H3I16w2D48w3a_2017_8"
+
+			# test APL compiler aliases and ignore set -e here
+			type xtensa-bxt-elf-gcc > /dev/null 2>&1 && true
```
Comment (marc-hb) [inline scripts/xtensa-build-all.sh:134]: I'm pretty sure `A && true` is the same than `A`

File: scripts/xtensa-build-all.sh [hunk 5]
```diff
@@ -113,171 +90,148 @@ set -e
...
-
-		# test APL compiler aliases and ignore set -e here
-		type xtensa-bxt-elf-gcc > /dev/null 2>&1 && true
-		if [ $? == 0 ]
-		then
-			HOST="xtensa-bxt-elf"
-		else
-			HOST="xtensa-apl-elf"
-		fi
+	case $platform in
```
Comment (marc-hb) [inline scripts/xtensa-build-all.sh:99]: These are associative arrays in disguise. Maybe we could add a dependency on bash version 4 in the future (not in this PR which is already fixing many other things).

File: scripts/xtensa-build-all.sh [hunk 6]
```diff
@@ -8,87 +8,64 @@ BUILD_ROM=no
...
-				then
-				BUILD_JOBS=$args
-				BUILD_JOBS_NEXT=0
-				continue
-			fi
-
-			printf '%s: WARN: ignoring arg %s\n' "$0" "$args" >&2
+		if [ $i == $arg ]
+		then
+			PLATFORMS+=$i" "
```
Comment (marc-hb) [inline scripts/xtensa-build-all.sh:57]: That's not how you expand arrays, I'm surprised this works...

https://mywiki.wooledge.org/BashGuide/Arrays

File: scripts/xtensa-build-all.sh [hunk 7]
```diff
@@ -8,87 +8,64 @@ BUILD_ROM=no
...
+    cat <<EOF
+usage: xtensa-build.sh [options] platform(s)
+       [-r] Build rom (gcc only)
+       [-a] Build all platforms
+       [-u] Force UP ARCH
+       [-d] Enable debug build
+       [-c] Configure defconfig
+       [-k] Use private key
+       [-v] Verbose Makefile log
+       [-j n] Set number of make build jobs. Jobs=#cores when no flag. \
```
[the remark this answers is not in the document]
Comment (meetmangukiya) [reply scripts/xtensa-build-all.sh:28]: It is an optional option, but requires a value if the option is specified. 
```
Usage: ping [-aAbBdDfhLnOqrRUvV64] [-c count] [-i interval] [-I interface]
```
Example output from `ping`'s help

[the remark this answers is not in the document]
Comment (crazoes) [reply scripts/xtensa-build-all.sh:28]: Yeah, it is an optional option, but if specified, need a value for that option. 
```
Usage: ping [-aAbBdDfhLnOqrRUvV64] [-c count] [-i interval] [-I interface]
```
`ping`'s help shows a similar option ^^

File: scripts/xtensa-build-all.sh (hunk 4 above)
Comment (marc-hb) [reply scripts/xtensa-build-all.sh:134]: `A || true` ignores `set -e` but it does that by making `$?` always true which _is_ obviously not wanted here considering the next line checks `$?`

The previous code didn't work with `set -e`. It worked without `set -e` but only because `&& true` did nothing, ironic. Can you tig blame these lines and find what happened? May also help find a way to test them.

`if type -p xtensa-bxt-elf-gcc; then...` is the only way to be compatible with `set -e`.

File: scripts/xtensa-build-all.sh (hunk 6 above)
Comment (marc-hb) [reply scripts/xtensa-build-all.sh:57]: Wow, I got to the bottom of this PLATFORMS thing and it's pretty bad. It's technically unrelated to the changes you're doing but I'm unfortunately not OK refactoring and moving around code that bad, even when you're not making it worse. Whoever wrote `PLATFORMS=${SUPPORTED_PLATFORMS[@]}` expected `PLATFORMS` to be an array. It's not, it's just one  string with spaces. This is ironic considering the main purpose of bash arrays is to manage list of strings with spaces in each; I digress.

Of course `PLATFORMS+=$i" "` is even "less" an array in this `else` clause. Maybe the author didn't really understand that arrays and strings are different types.

Yet there are references to PLATFORMS[@] below that seem to work [by coincidence](https://pragprog.com/the-pragmatic-programmer/extracts/coincidence).

File: scripts/xtensa-build-all.sh [hunk 8]
```diff
@@ -113,171 +90,148 @@ set -e
 OLDPATH=$PATH
 WORKDIR="$pwd"
 
-# build platform
-for j in ${PLATFORMS[@]}
+# build platforms
+for platform in ${PLATFORMS[@]}
```
Comment (marc-hb) [inline scripts/xtensa-build-all.sh:94]: It's a miracle this works considering PLATFORMS is not an array. This works by combining two bugs that cancel each other out:
- To list each element of the array it should be `for platform in "${PLATFORMS[@]}"` with double quotes
- but PLATFORMS is not an array but a string with whitespace

https://mywiki.wooledge.org/BashGuide/Arrays
Demo:
```diff
--- a/scripts/xtensa-build-all.sh
+++ b/scripts/xtensa-build-all.sh
@@ -90,6 +90,19 @@ else
        done
 fi
 
+# Here's an actual array
+declare -p SUPPORTED_PLATFORMS
+printf 'SPi=%s\n' "${SUPPORTED_PLATFORMS[@]}"
+
+# Not an array, just pretending
+declare -p PLATFORMS
+# yet bash expands this a string with whitespace, amazing
+printf 'Pi=%s\n' "${PLATFORMS[@]}"
+printf 'Pi=%s\n' "${PLATFORMS[@]}"
+
+# now something even weirder
+printf '?#?=%s\n' "${#PLATFORMS[@]}"
+
 # check target platform(s) have been passed in
 if [ ${#PLATFORMS[@]} -eq 0 ];
 then
@@ -97,6 +110,8 @@ then
        exit 1
 fi
 
+exit 0
+
 if [ "x$USE_PRIVATE_KEY" == "xyes" ]

```

File: scripts/xtensa-build-all.sh (hunk 6 above)
Comment (marc-hb) [reply scripts/xtensa-build-all.sh:57]: BTW I just saw the same array/string confusion in `./scripts/qemu-check.sh`

File: scripts/xtensa-build-all.sh (hunk 2 above)
Comment (dbaluta) [reply scripts/xtensa-build-all.sh:9]: @marc-hb I feel that we need to add a documentation chapter to SOF with how to contribute. We've seen people with different of styles and it would really. 

I'll add this on my TODO list for the next month. If anyone has some free time please jump on this.

File: scripts/xtensa-build-all.sh [hunk 9]
```diff
@@ -8,87 +8,64 @@ BUILD_ROM=no
...
+usage: xtensa-build.sh [options] platform(s)
+       [-r] Build rom (gcc only)
+       [-a] Build all platforms
+       [-u] Force UP ARCH
+       [-d] Enable debug build
+       [-c] Configure defconfig
+       [-k] Use private key
+       [-v] Verbose Makefile log
+       [-j n] Set number of make build jobs. Jobs=#cores when no flag. \
+Inifinte when not specified.
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply scripts/xtensa-build-all.sh:29]: I have also seen this typo, but since we are just now refactoring the old code this can be done in a separate PR. But now that you pointed it I think we can add a separate patch in this PR.
@crazoes Remeber that in a PR like this fixes come first, then the actual refactoring.

[the remark this answers is not in the document]
Comment (marc-hb) [reply scripts/xtensa-build-all.sh:29]: I like to treat typos the same as indentation/whitespace: don't obscure actual changes with them. However if that line changes ANYWAY in git for some other and good reason then it's a good opportunity to fix typos at the same time.

File: scripts/xtensa-build-all.sh (hunk 4 above)
Comment (marc-hb) [reply scripts/xtensa-build-all.sh:134]: There's a longer `set -e` story in issue #2752

Comment (crazoes) [reply scripts/xtensa-build-all.sh:134]: @marc-hb i see that you have exposed many major issues here, I will be addressing and working on these over the weekend

Comment (marc-hb) [reply scripts/xtensa-build-all.sh:134]: Thanks! Please don't get wrong: you probably do not want to fix every pre-existing issue that has been mentioned everywhere it can be found in the script. That would be a "Big Bang integration". As a rule of thumb you should fix pre-existing issues only in the code you're changing or moving, ideally in a separate commit before changing it or moving it.

Few hard and fast rules, lots of trade-offs.

File: scripts/xtensa-build-all.sh (hunk 5 above)
Comment (crazoes) [reply scripts/xtensa-build-all.sh:99]: Yeah, I just checked out associative arrays, this whole thing can be made into a nested associative array, with keys platform names, and values as another associative array containing `PLATFORM`, `ARCH`, etc..

File: scripts/xtensa-build-all.sh (hunk 3 above)
Comment (crazoes) [reply scripts/xtensa-build-all.sh:135]: If we don't redirect we end up with the path of `xtensa-bxt-elf-gcc` somewhere in the output?

File: scripts/xtensa-build-all.sh (hunk 4 above)
Comment (crazoes) [reply scripts/xtensa-build-all.sh:134]: > I'm pretty sure `A && true` is the same than `A`

Bash doesn't exit when `A && true` would be used, but it will exit if `A` alone is used. Found this answer: https://stackoverflow.com/questions/37838273/evaluating-bash-exit-codes-behaviour. And as you said `|| true` would execute when failed and change the exit code. 



> The previous code didn't work with `set -e`. It worked without `set -e` but only because `&& true` did nothing, ironic. Can you tig blame these lines and find what happened? May also help find a way to test them.

Not sure what you mean by previous code, because this code works with `set -e`, it is already set in the script before this command is executed. Also, here's the commit introducing this https://github.com/thesofproject/sof/commit/cad86dc340447f8ae514233d50c921a21c989d6b

Also, why are we adding the `-p` option? On checking the help, it seems like it will ignore if it is an alias. Not sure if anyone would alias this, but should we be ignoring aliases?

File: scripts/xtensa-build-all.sh (hunk 3 above)
Comment (marc-hb) [reply scripts/xtensa-build-all.sh:135]: Yes and I think it's useful information, think someone testing multiple toolchains version or a developer trying to make sense of an unknown docker image in CI. For someone who has only one toolchain this is just one extra line in an otherwise quite verbose log already. Either way, no big deal.

I've almost never encountered a situation where `/dev/null` is actually useful in a build script, I think it's just a really bad habit IMHO. I've seen `/dev/null` used even when the command has a `--quiet` option... Some tools (not `type`) don't use stderr, they print errors on stdout too then good luck when something goes wrong. If something is really too noisy and can't or shouldn't be turned off it should be sent to a log file or to a lower log level.

To give a bit more perspective: some CI engineers run almost all their CI scripts with `set -x`. They've lost too much time in the past trying to figure out some subtle change/difference in some random OS image.

File: scripts/xtensa-build-all.sh (hunk 4 above)
Comment (marc-hb) [reply scripts/xtensa-build-all.sh:134]: My bad about the `&& true`, thanks a lot for the explanation. I never can remember all the `set -e` exceptions, I wish this weren't one of them.

Please don't search and replace the whole file but in the bits you refactor please replace `&& true` with `|| true` anyway because `A || true`  reads as plain English, the `true` may actually run sometimes  and `|| true` would work even if that situation wasn't excluded by `set -e`, I mean `|| true` doesn't rely on some language subtlety like no-op `&& true` does.

The `-p` in `type -p` is not needed here, sorry for the noise, forget it. It's just the force of habit in: `foo=$(type -p bar)`. I forgot `-p` omits aliases and functions, thanks (that's one of the perks of reviewing code: learning goes both ways :-)  I never needed to test the existence of a function or (God forbid) an alias in a script.

**Comment by crazoes** [comment]
I don't know why but https://github.com/thesofproject/sof/pull/2711/commits doesn't show the right order of commit, rather ordered by dates. Making it difficult to read the actual sequence of the commits. I have checked the reviews again and verified I haven't missed any comment. However, there have been many amending and merge conflicts here so please let me know if I missed anything or if something might have lost due to wrongful merge conflict resolution on my part.

File: scripts/xtensa-build-all.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (crazoes) [reply scripts/xtensa-build-all.sh]: @marc-hb @dbaluta I have maintained the order here, fixing first and then moving the code around. 

In the process I have noticed that when we order moving of code after fixes, we will always end up with merge conflicts after editing the fix commit. So what is the rationale behind ordering fixes first and refactors later?

File: scripts/xtensa-build-all.sh [hunk 10]
```diff
@@ -113,171 +89,142 @@ set -e
...
+			HOST="xtensa-hsw-elf"
+			XTENSA_TOOLS_VERSION="RG-2017.8-linux"
+			;;
+		apl)
+			PLATFORM="apollolake"
+			ARCH="xtensa-smp"
+			XTENSA_CORE="X4H3I16w2D48w3a_2017_8"
+
+			# test APL compiler aliases and ignore set -e here
+			if ! type xtensa-bxt-elf-gcc; then
```
Comment (marc-hb) [inline scripts/xtensa-build-all.sh:133]: I think you meant: `if type xtensa-bxt-elf-gcc; then`.
This explains the Travis failure.

**Comment by marc-hb** [comment]
> I don't know why but https://github.com/thesofproject/sof/pull/2711/commits doesn't show the right order of commit, rather ordered by dates. Making it difficult to read the actual sequence of the commits. 
> I have checked the reviews again and verified I haven't missed any comment. However, there have been many amending and merge conflicts here so please let me know if I missed anything or if something might have lost due to wrongful merge conflict resolution on my part.

Wow, I think you discovered a new github limitation with force pushes, thanks! I added it to the list:
https://github.com/zephyrproject-rtos/zephyr/pull/14444#issuecomment-575320941


> In the process I have noticed that when we order moving of code after fixes, we will always end up with merge conflicts after editing the fix commit. So what is the rationale behind ordering fixes first and refactors later?

Yes, re-ordering commits that touch the same parts of the code will always cause painful conflicts. The rationale is that version history is sometimes painfully written once by one person (you), but then it's immutable and read, bisected, backported and used in many other ways by many other people and many tools. This ["clean history"](https://lwn.net/Articles/328436/) rationale is a very common rationale among git users, apparently much less common among [github users](https://github.com/isaacs/github/issues/959#issuecomment-5929784740), github seems to care less about history.

> ordering fixes first and refactors later?

It's not always like this, it depends. In other cases a zero-functional change refactor is a prelude to shorten and clarify a bug fix coming immediately next and making the fix much easier to revert in case it regresses something else. In this PR the hopefully small fix(es) were completely unrelated to the refactor, so better not increase and obscure the distance between the bugs and their fixes by shuffling code in between.

PS: besides root-causing the Travis failure, I haven't reviewed the latest version yet.

File: scripts/xtensa-build-all.sh (hunk 10 above)
Comment (crazoes) [reply scripts/xtensa-build-all.sh:133]: Either way `HOST` is never set to `xtensa-bxt-elf-gcc` which is what is being set as the compiler. Currently unable to get the docker image because of network issues so cannot reproduce locally. Anyways, I will try removing the `!` and see if it works
```
  The CMAKE_C_COMPILER:

    xtensa-bxt-elf-gcc

  is not a full path and was not found in the PATH.

```

Comment (crazoes) [reply scripts/xtensa-build-all.sh]: huh, this does fix it. But doesn't work well with what I said earlier. Maybe the `CMAKE_C_COMPILER` is not derived from `TOOLCHAIN` which is what is using the `HOST` value?

Comment (marc-hb) [reply scripts/xtensa-build-all.sh]: `bxt` and `apl` were technically different but have been used somewhat interchangeably in some contexts. If you look at commit c7edd736d9c6 you can see this logic is only backward compatibility to support who built a toolchain with the older `bxt` name. All documentation now refers to `apl`

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by xiulipan** [comment]
@crazoes What the change about the `-j` option? Why you would like to remove it now?

**Comment by crazoes** [comment]
> @crazoes What the change about the `-j` option? Why you would like to remove it now?

@xiulipan Explained in another discussion here : https://github.com/thesofproject/sof/pull/2711#discussion_r403507570

**Comment by xiulipan** [comment]
@crazoes I will try to remove the -j option in our Jenkins CI scripts.

**Comment by lgirdwood** [comment]
CI known issues.

**Comment by marc-hb** [comment]
This PR missed one `$j` -> `$platform` replacement:
```
$ shellcheck scripts/xtensa-build-all.sh 
In scripts/xtensa-build-all.sh line 273:
```
```shell
       case $platform in
       ...

		TOOLCHAIN=$HOST
		PATH=$pwd/../$HOST/bin:$OLDPATH
		COMPILER="gcc"

		case $j in
                     ^-- SC2154: j is referenced but not assigned.

			byt|cht|cnl|sue) DEFCONFIG_PATCH="_gcc";;
			*)	     DEFCONFIG_PATCH="";;
		esac
```

I think this means the wrong `_defconfig` files have been used for `byt|cht|cnl|sue` for the last 3- months?

(unrelated but large shellcheck noise reduced in unrelated PR #3163 )

**Comment by crazoes** [comment]
@marc-hb you are right. I will send a fix.
